File: src/nostr/wallet_registry.rs
use std::collections::{HashMap, HashSet};
use async_trait::async_trait;
use serde_json::Value;
use crate::log_debug;
use crate::log_info;
use crate::log_error;
use crate::nostr::{Filter, Event, Limits};
use crate::util::short;

#[async_trait(?Send)]
pub trait Storage {
    async fn get(&self, key: &str) -> Option<Value>;
    async fn put_batch(&self, entries: HashMap<String, Value>) -> Result<(), String>;
    async fn delete_batch(&self, keys: Vec<String>);
}

pub struct SavedState {
    pub json: Value,
    pub pubkeys: HashSet<String>,
}

/// A purely synchronous index for subscriptions and info events.
type PkEntry = (HashSet<(String, Vec<Filter>)>, Option<Event>);

struct WalletIndex {
    subscription_index: HashMap<(String, Vec<Filter>), Vec<u32>>,
    pk_index: HashMap<String, PkEntry>,
    info_id_index: HashMap<String, String>,
    reverse_index: HashMap<u32, HashMap<String, Vec<Filter>>>,
}

impl WalletIndex {
    fn new() -> Self {
        Self {
            subscription_index: HashMap::new(),
            pk_index: HashMap::new(),
            info_id_index: HashMap::new(),
            reverse_index: HashMap::new(),
        }
    }

    fn subscribe(&mut self, conn_id: u32, sub_id: String, filters: Vec<Filter>) {
        self.unsubscribe(conn_id, sub_id.clone());

        let sub_key = (sub_id.clone(), filters.clone());

        for filter in &filters {
            for pk in filter.pubkeys() {
                let entry = self.pk_index.entry(pk).or_insert_with(|| (HashSet::new(), None));
                entry.0.insert(sub_key.clone());
            }
        }

        let conns = self.subscription_index.entry(sub_key).or_default();
        if !conns.contains(&conn_id) {
            conns.push(conn_id);
        }

        self.reverse_index.entry(conn_id)
            .or_default()
            .insert(sub_id, filters);
    }

    fn unsubscribe(&mut self, conn_id: u32, sub_id: String) {
        if let Some(conn_subs) = self.reverse_index.get_mut(&conn_id) {
            if let Some(filters) = conn_subs.remove(&sub_id) {
                let sub_key = (sub_id, filters);

                if let Some(conns) = self.subscription_index.get_mut(&sub_key) {
                    conns.retain(|&id| id != conn_id);
                    if conns.is_empty() {
                        self.subscription_index.remove(&sub_key);

                        for filter in &sub_key.1 {
                            for pk in filter.pubkeys() {
                                if let Some(entry) = self.pk_index.get_mut(&pk) {
                                    entry.0.remove(&sub_key);
                                    if entry.0.is_empty() && entry.1.is_none() {
                                        self.pk_index.remove(&pk);
                                    }
                                }
                            }
                        }
                    }
                }
            }
            if conn_subs.is_empty() {
                self.reverse_index.remove(&conn_id);
            }
        }
    }

    fn disconnect(&mut self, conn_id: u32) {
        let sub_ids: Vec<String> = self.reverse_index.get(&conn_id)
            .map(|m| m.keys().cloned().collect())
            .unwrap_or_default();

        for sub_id in sub_ids {
            self.unsubscribe(conn_id, sub_id);
        }
    }

    fn get_subscriptions(&self, conn_id: u32) -> HashMap<String, Vec<Filter>> {
        self.reverse_index.get(&conn_id).cloned().unwrap_or_default()
    }

    fn sub_count(&self, conn_id: u32) -> usize {
        self.reverse_index.get(&conn_id).map_or(0, |m| m.len())
    }

    fn cache_info(&mut self, event: Event) {
        let entry = self.pk_index.entry(event.pubkey.clone()).or_insert_with(|| (HashSet::new(), None));
        self.info_id_index.insert(event.id.clone(), event.pubkey.clone());
        entry.1 = Some(event);
    }

    fn get_info(&self, pubkey: &str) -> Option<Event> {
        self.pk_index.get(pubkey).and_then(|e| e.1.clone())
    }

    fn delete_info(&mut self, pubkey: &str) -> Option<Event> {
        let removed = if let Some(entry) = self.pk_index.get_mut(pubkey) {
            let old = entry.1.take();
            if let Some(ref event) = old {
                self.info_id_index.remove(&event.id);
            }
            old
        } else {
            None
        };
        // Clean up pk_index entry if both subscriptions and info are gone
        if let Some(entry) = self.pk_index.get(pubkey) {
            if entry.0.is_empty() && entry.1.is_none() {
                self.pk_index.remove(pubkey);
            }
        }
        removed
    }

    fn find_info_pubkey_by_id(&self, event_id: &str) -> Option<String> {
        self.info_id_index.get(event_id).cloned()
    }

    fn match_event<'a>(&'a self, event: &'a Event) -> Box<dyn Iterator<Item = (String, Vec<u32>)> + 'a> {
        let target_pks = event.target_pubkeys();

        let mut sub_to_conns = HashMap::new();
        for pk in target_pks {
            if let Some(entry) = self.pk_index.get(&pk) {
                let mut matching: Vec<(String, u32)> = Vec::new();
                for sub_key in &entry.0 {
                    if sub_key.1.iter().any(|f| f.matches(event)) {
                        if let Some(conns) = self.subscription_index.get(sub_key) {
                            if let Some(&latest_conn) = conns.last() {
                                matching.push((sub_key.0.clone(), latest_conn));
                            }
                        }
                    }
                }
                let latest = matching.iter().map(|(_, id)| *id).max();
                if let Some(latest_id) = latest {
                    for (sub_id, id) in matching {
                        if id == latest_id {
                            sub_to_conns.insert(sub_id, vec![id]);
                        }
                    }
                }
            }
        }
        Box::new(sub_to_conns.into_iter())
    }

    fn get_connection_id(&self, pubkey: &str) -> Option<u32> {
        if let Some(entry) = self.pk_index.get(pubkey) {
            let mut latest = None;
            for sub_key in &entry.0 {
                if let Some(conns) = self.subscription_index.get(sub_key) {
                    if let Some(&id) = conns.last() {
                        match latest {
                            None => latest = Some(id),
                            Some(current) if id > current => latest = Some(id),
                            _ => {}
                        }
                    }
                }
            }
            return latest;
        }
        None
    }

    fn save(&self, conn_id: u32) -> Option<SavedState> {
        let subscriptions = self.get_subscriptions(conn_id);
        if subscriptions.is_empty() {
            return None;
        }

        let mut info_event = None;
        let mut pubkeys = HashSet::new();
        for filters in subscriptions.values() {
            for filter in filters {
                for pk in filter.pubkeys() {
                    pubkeys.insert(pk.clone());
                    if info_event.is_none() {
                        if let Some(event) = self.get_info(&pk) {
                            info_event = Some(event);
                        }
                    }
                }
            }
        }

        Some(SavedState {
            json: serde_json::json!({
                "subscriptions": subscriptions,
                "info_event": info_event,
            }),
            pubkeys,
        })
    }

    fn restore(&mut self, conn_id: u32, data: Value) {
        if let Some(subs_val) = data.get("subscriptions") {
            if let Ok(subs) = serde_json::from_value::<HashMap<String, Vec<Filter>>>(subs_val.clone()) {
                for (sub_id, filters) in subs {
                    self.subscribe(conn_id, sub_id, filters);
                }
            }
        }
        if let Some(info_val) = data.get("info_event") {
            if let Ok(event) = serde_json::from_value::<Event>(info_val.clone()) {
                self.cache_info(event);
            }
        }
    }
}

pub struct WalletRegistry<S: Storage> {
    storage: S,
    index: WalletIndex,
    limits: Limits,
}

#[derive(Debug, PartialEq, Eq)]
pub enum RegistryResponse {
    Send { recipient_id: u32, sub_id: String },
    WakeUp(u32),
}

impl<S: Storage> WalletRegistry<S> {
    pub fn new(storage: S, limits: Limits) -> Self {
        Self { 
            storage,
            index: WalletIndex::new(),
            limits,
        }
    }

    pub async fn subscribe(&mut self, conn_id: u32, sub_id: String, filters: Vec<Filter>) -> crate::nostr::Result<()> {
        let count = self.index.sub_count(conn_id);
        if count >= self.limits.max_subscriptions_per_connection {
            return Err(crate::nostr::RelayError::Generic(format!(
                "too many subscriptions: {} (max {})",
                count, self.limits.max_subscriptions_per_connection
            )));
        }
        self.index.subscribe(conn_id, sub_id, filters);
        self.sync(conn_id).await.map_err(|e| crate::nostr::RelayError::Generic(format!("persist failed: {}", e)))?;
        Ok(())
    }

    pub async fn unsubscribe(&mut self, conn_id: u32, sub_id: String) -> crate::nostr::Result<()> {
        self.index.unsubscribe(conn_id, sub_id.clone());
        if let Err(e) = self.sync(conn_id).await {
            log_error!("unsubscribe sync failed: conn={} sub={} err={}", conn_id, sub_id, e);
        }
        Ok(())
    }

    pub async fn match_event(&mut self, event: &Event) -> Vec<RegistryResponse> {
        let mut responses = Vec::new();
        let target_pks = event.target_pubkeys();

        for pk in target_pks {
            for id in self.load_by_pubkey(&pk).await {
                responses.push(RegistryResponse::WakeUp(id));
            }
        }

        for (sub_id, conns) in self.index.match_event(event) {
            for id in conns {
                responses.push(RegistryResponse::Send { recipient_id: id, sub_id: sub_id.clone() });
            }
        }
        responses
    }

    pub async fn load(&mut self, conn_id: u32) -> bool {
        if !self.index.get_subscriptions(conn_id).is_empty() {
            return true;
        }
        let key = format!("conn:{}", conn_id);
        if let Some(data) = self.storage.get(&key).await {
            self.index.restore(conn_id, data);
            let count = self.index.get_subscriptions(conn_id).len();
            log_info!("restored conn={} with {} subscriptions", conn_id, count);
            return true;
        }
        log_debug!("load: conn={} not found in storage", conn_id);
        false
    }

    pub async fn load_by_pubkey(&mut self, pubkey: &str) -> Vec<u32> {
        let key = format!("pk:{}", pubkey);
        let storage_ids: Vec<u32> = if let Some(val) = self.storage.get(&key).await {
            match &val {
                Value::Array(arr) => arr.iter().filter_map(|v| v.as_u64().map(|x| x as u32)).collect(),
                Value::Number(n) => n.as_u64().map(|x| vec![x as u32]).unwrap_or_default(),
                _ => Vec::new(),
            }
        } else {
            if let Some(id) = self.index.get_connection_id(pubkey) {
                return vec![id];
            }
            return Vec::new();
        };
        let mut loaded = Vec::new();
        let mut stale = Vec::new();
        for id in storage_ids {
            if self.load(id).await {
                loaded.push(id);
            } else {
                stale.push(id);
            }
        }
        if loaded.is_empty() {
            log_debug!("stale pubkey index cleaned: pk={}", short(pubkey, 8));
            self.storage.delete_batch(vec![key]).await;
        } else if !stale.is_empty() {
            let mut entries = HashMap::new();
            entries.insert(key, serde_json::json!(loaded));
            let _ = self.storage.put_batch(entries).await;
        }
        loaded
    }

    pub async fn on_disconnect(&mut self, id: u32) {
        self.index.disconnect(id);
    }

    pub async fn on_terminate(&mut self, id: u32) {
        let subs = self.index.get_subscriptions(id);
        let mut pubkeys: HashSet<String> = HashSet::new();
        for filters in subs.values() {
            for filter in filters {
                for pk in filter.pubkeys() {
                    pubkeys.insert(pk);
                }
            }
        }

        self.index.disconnect(id);

        for pk in &pubkeys {
            let key = format!("pk:{}", pk);
            if let Some(val) = self.storage.get(&key).await {
                let ids: Vec<u32> = match &val {
                    Value::Array(arr) => arr.iter().filter_map(|v| v.as_u64().map(|x| x as u32)).collect(),
                    Value::Number(n) => n.as_u64().map(|x| vec![x as u32]).unwrap_or_default(),
                    _ => Vec::new(),
                };
                let new_ids: Vec<u32> = ids.into_iter().filter(|x| *x != id).collect();
                if new_ids.is_empty() {
                    self.storage.delete_batch(vec![key]).await;
                } else {
                    let mut entries = HashMap::new();
                    entries.insert(key, serde_json::json!(new_ids));
                    let _ = self.storage.put_batch(entries).await;
                }
            }
        }

        log_debug!("deleted conn state: conn={}", id);
        self.storage.delete_batch(vec![format!("conn:{}", id)]).await;
    }

    pub async fn cache_info(&mut self, event: Event) {
        log_debug!("persist info: pk={}", short(&event.pubkey, 8));
        let key = format!("info:{}", event.pubkey);
        let value = serde_json::to_value(&event).unwrap_or(serde_json::Value::Null);
        if !value.is_null() {
            let mut entries = HashMap::new();
            entries.insert(key, value);
            let _ = self.storage.put_batch(entries).await;
        }
        self.index.cache_info(event);
    }

    pub async fn get_info(&mut self, pubkey: &str) -> Option<Event> {
        if let Some(event) = self.index.get_info(pubkey) {
            return Some(event);
        }

        let key = format!("info:{}", pubkey);
        if let Some(val) = self.storage.get(&key).await {
            if let Ok(event) = serde_json::from_value::<Event>(val) {
                log_debug!("info restored from storage: pk={}", short(pubkey, 8));
                self.index.cache_info(event.clone());
                return Some(event);
            }
        }
        None
    }

    pub async fn delete_info(&mut self, pubkey: &str) {
        let removed = self.index.delete_info(pubkey);
        if removed.is_some() {
            log_debug!("info deleted: pk={}", short(pubkey, 8));
            let key = format!("info:{}", pubkey);
            self.storage.delete_batch(vec![key]).await;
        }
    }

    pub fn find_info_pubkey_by_id(&self, event_id: &str) -> Option<String> {
        self.index.find_info_pubkey_by_id(event_id)
    }

    async fn sync(&self, conn_id: u32) -> Result<(), String> {
        if let Some(state) = self.index.save(conn_id) {
            let mut entries = HashMap::new();
            entries.insert(format!("conn:{}", conn_id), state.json);
            for pk in state.pubkeys {
                let key = format!("pk:{}", pk);
                let mut ids = self.read_pk_list(&key).await;
                if !ids.contains(&conn_id) {
                    ids.push(conn_id);
                }
                entries.insert(key, serde_json::json!(ids));
            }
            self.storage.put_batch(entries).await?;
        } else {
            self.storage.delete_batch(vec![format!("conn:{}", conn_id)]).await;
        }
        Ok(())
    }

    async fn read_pk_list(&self, key: &str) -> Vec<u32> {
        if let Some(val) = self.storage.get(key).await {
            match &val {
                Value::Array(arr) => arr.iter().filter_map(|v| v.as_u64().map(|x| x as u32)).collect(),
                Value::Number(n) => n.as_u64().map(|x| vec![x as u32]).unwrap_or_default(),
                _ => Vec::new(),
            }
        } else {
            Vec::new()
        }
    }
}

#[cfg(test)]
#[path = "wallet_registry_test.rs"]
mod wallet_registry_test;

File: src/nostr/wallet_registry.ts
import type { Event } from './event.ts';
import type { Filter } from './filter.ts';
import type { Limits } from './limits.ts';
import { filterMatches, filterPubkeys } from './filter.ts';
import { targetPubkeys } from './event.ts';
import { RelayError } from './error.ts';

export interface Storage {
  get(key: string): Promise<unknown | null>;
  putBatch(entries: Record<string, unknown>): Promise<void>;
  deleteBatch(keys: string[]): Promise<void>;
}

export interface SavedState {
  json: unknown;
  pubkeys: string[];
}

export type RegistryResponse =
  | { kind: 'send'; recipientId: number; subId: string }
  | { kind: 'wakeUp'; connectionId: number };

type PkEntry = { subs: Set<string>; info: Event | null };

function subKey(subId: string, filters: Filter[]): string {
  return subId + '::' + JSON.stringify(filters);
}

class WalletIndex {
  private subscriptionIndex = new Map<string, number[]>();
  private pkIndex = new Map<string, PkEntry>();
  private infoIdIndex = new Map<string, string>();
  private reverseIndex = new Map<number, Map<string, Filter[]>>();

  subscribe(connId: number, subId: string, filters: Filter[]): void {
    this.unsubscribe(connId, subId);

    const key = subKey(subId, filters);

    for (const filter of filters) {
      for (const pk of filterPubkeys(filter)) {
        let entry = this.pkIndex.get(pk);
        if (!entry) {
          entry = { subs: new Set(), info: null };
          this.pkIndex.set(pk, entry);
        }
        entry.subs.add(key);
      }
    }

    let conns = this.subscriptionIndex.get(key);
    if (!conns) {
      conns = [];
      this.subscriptionIndex.set(key, conns);
    }
    if (!conns.includes(connId)) {
      conns.push(connId);
    }

    let connSubs = this.reverseIndex.get(connId);
    if (!connSubs) {
      connSubs = new Map();
      this.reverseIndex.set(connId, connSubs);
    }
    connSubs.set(subId, filters);
  }

  unsubscribe(connId: number, subId: string): void {
    const connSubs = this.reverseIndex.get(connId);
    if (!connSubs) return;

    const filters = connSubs.get(subId);
    if (filters === undefined) return;
    connSubs.delete(subId);

    const key = subKey(subId, filters);
    const conns = this.subscriptionIndex.get(key);
    if (conns) {
      const idx = conns.indexOf(connId);
      if (idx !== -1) conns.splice(idx, 1);
      if (conns.length === 0) {
        this.subscriptionIndex.delete(key);
        for (const filter of filters) {
          for (const pk of filterPubkeys(filter)) {
            const entry = this.pkIndex.get(pk);
            if (entry) {
              entry.subs.delete(key);
              if (entry.subs.size === 0 && entry.info === null) {
                this.pkIndex.delete(pk);
              }
            }
          }
        }
      }
    }

    if (connSubs.size === 0) {
      this.reverseIndex.delete(connId);
    }
  }

  disconnect(connId: number): void {
    const connSubs = this.reverseIndex.get(connId);
    if (!connSubs) return;
    for (const subId of connSubs.keys()) {
      this.unsubscribe(connId, subId);
    }
  }

  getSubscriptions(connId: number): Map<string, Filter[]> {
    return new Map(this.reverseIndex.get(connId) ?? new Map());
  }

  subCount(connId: number): number {
    return this.reverseIndex.get(connId)?.size ?? 0;
  }

  cacheInfo(event: Event): void {
    let entry = this.pkIndex.get(event.pubkey);
    if (!entry) {
      entry = { subs: new Set(), info: null };
      this.pkIndex.set(event.pubkey, entry);
    }
    this.infoIdIndex.set(event.id, event.pubkey);
    entry.info = event;
  }

  getInfo(pubkey: string): Event | null {
    return this.pkIndex.get(pubkey)?.info ?? null;
  }

  deleteInfo(pubkey: string): Event | null {
    const entry = this.pkIndex.get(pubkey);
    if (!entry) return null;

    const old = entry.info;
    if (old) {
      this.infoIdIndex.delete(old.id);
    }
    entry.info = null;

    if (entry.subs.size === 0 && entry.info === null) {
      this.pkIndex.delete(pubkey);
    }

    return old;
  }

  findInfoPubkeyById(eventId: string): string | null {
    return this.infoIdIndex.get(eventId) ?? null;
  }

  matchEvent(event: Event): Map<string, number[]> {
    const targetPks = targetPubkeys(event);
    const subToConns = new Map<string, number[]>();

    for (const pk of targetPks) {
      const entry = this.pkIndex.get(pk);
      if (!entry) continue;

      const matching: Array<{ subId: string; connId: number }> = [];
      for (const key of entry.subs) {
        const sepIdx = key.indexOf('::');
        const subId = key.substring(0, sepIdx);
        const filtersJson = key.substring(sepIdx + 2);
        try {
          const filters = JSON.parse(filtersJson) as Filter[];
          if (filters.some(f => filterMatches(f, event))) {
            const conns = this.subscriptionIndex.get(key);
            if (conns && conns.length > 0) {
              const latestConn = conns[conns.length - 1];
              matching.push({ subId, connId: latestConn });
            }
          }
        } catch {
          // skip malformed filter JSON
        }
      }

      if (matching.length === 0) continue;
      const latestId = Math.max(...matching.map(m => m.connId));
      for (const m of matching) {
        if (m.connId === latestId) {
          subToConns.set(m.subId, [m.connId]);
        }
      }
    }

    return subToConns;
  }

  getConnectionId(pubkey: string): number | null {
    const entry = this.pkIndex.get(pubkey);
    if (!entry) return null;

    let latest: number | null = null;
    for (const key of entry.subs) {
      const conns = this.subscriptionIndex.get(key);
      if (conns && conns.length > 0) {
        const id = conns[conns.length - 1];
        if (latest === null || id > latest) {
          latest = id;
        }
      }
    }
    return latest;
  }

  save(connId: number): SavedState | null {
    const subscriptions = this.getSubscriptions(connId);
    if (subscriptions.size === 0) return null;

    let infoEvent: Event | null = null;
    const pubkeys: string[] = [];

    for (const filters of subscriptions.values()) {
      for (const filter of filters) {
        for (const pk of filterPubkeys(filter)) {
          if (!pubkeys.includes(pk)) pubkeys.push(pk);
          if (!infoEvent) {
            infoEvent = this.getInfo(pk);
          }
        }
      }
    }

    return {
      json: {
        subscriptions: Object.fromEntries(subscriptions),
        info_event: infoEvent,
      },
      pubkeys,
    };
  }

  restore(connId: number, data: Record<string, unknown>): void {
    const subsData = data.subscriptions;
    if (subsData && typeof subsData === 'object') {
      for (const [subId, filters] of Object.entries(subsData)) {
        if (Array.isArray(filters)) {
          this.subscribe(connId, subId, filters as Filter[]);
        }
      }
    }
    const infoData = data.info_event;
    if (infoData && typeof infoData === 'object') {
      this.cacheInfo(infoData as unknown as Event);
    }
  }
}

export class WalletRegistry<S extends Storage> {
  storage: S;
  private index = new WalletIndex();
  private limits: Limits;

  constructor(storage: S, limits: Limits) {
    this.storage = storage;
    this.limits = limits;
  }

  async subscribe(connId: number, subId: string, filters: Filter[]): Promise<void> {
    const count = this.index.subCount(connId);
    if (count >= this.limits.maxSubscriptionsPerConnection) {
      throw RelayError.Generic(
        `too many subscriptions: ${count} (max ${this.limits.maxSubscriptionsPerConnection})`,
      );
    }
    this.index.subscribe(connId, subId, filters);
    try {
      await this.sync(connId);
    } catch (e) {
      throw RelayError.Generic(`persist failed: ${(e as Error).message}`);
    }
  }

  async unsubscribe(connId: number, subId: string): Promise<void> {
    this.index.unsubscribe(connId, subId);
    try {
      await this.sync(connId);
    } catch (e) {
      // log error but don't propagate (graceful on storage failure)
    }
  }

  async matchEvent(event: Event): Promise<RegistryResponse[]> {
    const responses: RegistryResponse[] = [];
    const targetPks = targetPubkeys(event);

    for (const pk of targetPks) {
      for (const id of await this.loadByPubkey(pk)) {
        responses.push({ kind: 'wakeUp', connectionId: id });
      }
    }

    for (const [subId, conns] of this.index.matchEvent(event)) {
      for (const id of conns) {
        responses.push({ kind: 'send', recipientId: id, subId });
      }
    }

    return responses;
  }

  async load(connId: number): Promise<boolean> {
    if (this.index.getSubscriptions(connId).size > 0) return true;

    const key = `conn:${connId}`;
    const data = await this.storage.get(key);
    if (data && typeof data === 'object') {
      this.index.restore(connId, data as Record<string, unknown>);
      return true;
    }
    return false;
  }

  async loadByPubkey(pubkey: string): Promise<number[]> {
    const key = `pk:${pubkey}`;
    const val = await this.storage.get(key);

    let storageIds: number[];
    if (val !== null) {
      if (Array.isArray(val)) {
        storageIds = val.map(v => Number(v));
      } else if (typeof val === 'number') {
        storageIds = [val];
      } else {
        storageIds = [];
      }
    } else {
      const id = this.index.getConnectionId(pubkey);
      return id !== null ? [id] : [];
    }

    const loaded: number[] = [];
    const stale: number[] = [];
    for (const id of storageIds) {
      if (await this.load(id)) {
        loaded.push(id);
      } else {
        stale.push(id);
      }
    }

    if (loaded.length === 0) {
      await this.storage.deleteBatch([key]);
    } else if (stale.length > 0) {
      await this.storage.putBatch({ [key]: loaded });
    }

    return loaded;
  }

  async onDisconnect(id: number): Promise<void> {
    this.index.disconnect(id);
  }

  async onTerminate(id: number): Promise<void> {
    const subs = this.index.getSubscriptions(id);
    const pubkeys = new Set<string>();
    for (const filters of subs.values()) {
      for (const filter of filters) {
        for (const pk of filterPubkeys(filter)) {
          pubkeys.add(pk);
        }
      }
    }

    this.index.disconnect(id);

    for (const pk of pubkeys) {
      const key = `pk:${pk}`;
      const val = await this.storage.get(key);
      if (val !== null) {
        let ids: number[];
        if (Array.isArray(val)) {
          ids = val.map(v => Number(v));
        } else if (typeof val === 'number') {
          ids = [val];
        } else {
          ids = [];
        }
        const newIds = ids.filter(x => x !== id);
        if (newIds.length === 0) {
          await this.storage.deleteBatch([key]);
        } else {
          await this.storage.putBatch({ [key]: newIds });
        }
      }
    }

    await this.storage.deleteBatch([`conn:${id}`]);
  }

  async cacheInfo(event: Event): Promise<void> {
    const key = `info:${event.pubkey}`;
    await this.storage.putBatch({ [key]: event });
    this.index.cacheInfo(event);
  }

  async getInfo(pubkey: string): Promise<Event | null> {
    const cached = this.index.getInfo(pubkey);
    if (cached) return cached;

    const key = `info:${pubkey}`;
    const val = await this.storage.get(key);
    if (val && typeof val === 'object') {
      const event = val as Event;
      this.index.cacheInfo(event);
      return event;
    }
    return null;
  }

  async deleteInfo(pubkey: string): Promise<void> {
    const removed = this.index.deleteInfo(pubkey);
    if (removed) {
      const key = `info:${pubkey}`;
      await this.storage.deleteBatch([key]);
    }
  }

  findInfoPubkeyById(eventId: string): string | null {
    return this.index.findInfoPubkeyById(eventId);
  }

  // test helper: expose subscriptions for a connection
  getConnSubscriptions(connId: number): Map<string, Filter[]> {
    return this.index.getSubscriptions(connId);
  }

  private async sync(connId: number): Promise<void> {
    const state = this.index.save(connId);
    if (state) {
      const entries: Record<string, unknown> = {
        [`conn:${connId}`]: state.json,
      };
      for (const pk of state.pubkeys) {
        const key = `pk:${pk}`;
        const ids = await this.readPkList(key);
        if (!ids.includes(connId)) {
          ids.push(connId);
        }
        entries[key] = ids;
      }
      await this.storage.putBatch(entries);
    } else {
      await this.storage.deleteBatch([`conn:${connId}`]);
    }
  }

  private async readPkList(key: string): Promise<number[]> {
    const val = await this.storage.get(key);
    if (val !== null) {
      if (Array.isArray(val)) {
        return val.map(v => Number(v));
      } else if (typeof val === 'number') {
        return [val];
      }
    }
    return [];
  }
}
